Eliminate Gradle boilerplate for pure JS projects

serhii-lekariev opened the pull request:
This PR, together with https://github.com/SpineEventEngine/base/pull/419, fixes https://github.com/SpineEventEngine/bootstrap/issues/16.

Before, pure JavaScript projects, i.e projects that only declare `spine.enableJavaScript()`, required a lot of boilerplate, namely:
* exclusion of transitive `protobuf` dependencies in the `configurations` block. Example:
```groovy
configurations {
    protobuf.exclude group: 'io.spine', module: 'spine-base'
    protobuf.exclude group: 'io.spine', module: 'spine-core'
    protobuf.exclude group: 'io.spine', module: 'spine-client'
    protobuf.exclude group: 'com.google.protobuf'
}
```
* `protoc` plugin configurations, such as whether the descriptor set is to be generated, the location of the descriptor set, etc. Example: 
```groovy
protobuf {
    generatedFilesBaseDir = compiledProto
    protoc {
        artifact = deps.build.protoc
    }
    generateProtoTasks {
        all().each { final task ->
            task.plugins {
                task.generateDescriptorSet = false
            }
            task.builtins {
                remove java
                js {
                    option "import_style=commonjs"
                }
                task.generateDescriptorSet = true
                final def testClassifier = task.sourceSet.name == "test" ? "_test" : ""
                final def descriptorName = "${project.group}_${project.name}_${project.version}${testClassifier}.desc"
                task.descriptorSetOptions.path = "${projectDir}/build/descriptors/${task.sourceSet.name}/${descriptorName}"
            }
            compileProtoToJs.dependsOn task
        }
    }
}
```
* disabling `compileJava` and `compileTestJava` tasks.

serhii-lekariev committed:
Fix generated JS files lookup in tests. Add a test for
`index.js` presence.

Also update the Spine version.

serhii-lekariev committed:
Merge branch 'master' into pure-js-boilerplate-elimination

serhii-lekariev committed:
Update config and add methods that toggle Gradle tasks exclusive to Java

serhii-lekariev committed:
Remove a redundant test case and add a new one.

serhii-lekariev committed:
Update README.md to reflect the fixes in https://github.com/SpineEventEngine/bootstrap/pull/18.

serhii-lekariev commented:
@dmdashenkov, PTAL.

dmdashenkov commented on plugin/src/main/java/io/spine/tools/gradle/bootstrap/Extension.java:
```diff
@@ -117,6 +136,51 @@ public JavaScriptExtension enableJavaScript() {
+        toggleJavaTasks(false);
+    }
+
+    private void toggleJavaTasks(boolean shouldEnable) {
```
```suggestion
    private void toggleJavaTasks(boolean enabled) {
```

dmdashenkov commented on plugin/src/main/java/io/spine/tools/gradle/bootstrap/Extension.java:
```diff
@@ -117,6 +136,51 @@ public JavaScriptExtension enableJavaScript() {
+     *
+     * @param shouldEnable whether the transitivity should be enabled
+     */
+    private void toggleTransitiveProtos(boolean shouldEnable) {
```
Same as above.

dmdashenkov commented on plugin/src/main/java/io/spine/tools/gradle/bootstrap/Extension.java:
```diff
@@ -117,6 +136,51 @@ public JavaScriptExtension enableJavaScript() {
+        project.configurations(closure((ConfigurationContainer container) -> {
+            try {
+                @SuppressWarnings("DuplicateStringLiteralInspection" /* Different context. */)
+                Configuration protobuf = container.getByName("protobuf");
```
There must be a constant in `CnfigurationName` in `base`. If it's not there, please add it.

dmdashenkov commented on plugin/src/main/java/io/spine/tools/gradle/bootstrap/Extension.java:
```diff
@@ -117,6 +136,51 @@ public JavaScriptExtension enableJavaScript() {
+            try {
+                @SuppressWarnings("DuplicateStringLiteralInspection" /* Different context. */)
+                Configuration protobuf = container.getByName("protobuf");
+                protobuf.setTransitive(shouldEnable);
```
`protobuf` configuration should be transitive regardless of Java/JavaScript.

dmdashenkov commented on plugin/src/main/java/io/spine/tools/gradle/bootstrap/Extension.java:
```diff
@@ -117,6 +136,51 @@ public JavaScriptExtension enableJavaScript() {
+                @SuppressWarnings("DuplicateStringLiteralInspection" /* Different context. */)
+                Configuration protobuf = container.getByName("protobuf");
+                protobuf.setTransitive(shouldEnable);
+            } catch (UnknownConfigurationException ignored) {
```
There must be a `find` method, which returns `null` instead of throwing an exception.

dmdashenkov commented on plugin/src/main/java/io/spine/tools/gradle/bootstrap/Extension.java:
```diff
@@ -117,6 +136,51 @@ public JavaScriptExtension enableJavaScript() {
+        Optional.ofNullable(tasks.findByPath(COMPILE_JAVA))
+                .ifPresent(task -> task.setEnabled(shouldEnable));
+        Optional.ofNullable(tasks.findByPath(COMPILE_TEST_JAVA))
+                .ifPresent(task -> task.setEnabled(shouldEnable));
```
Please use `if (<..> != null)` constructs instead. Using `Optional` locally for branching is considered a bad practice.

dmdashenkov reviewed (changes requested):
@serhii-lekariev, please see my comments.

serhii-lekariev committed:
Disable transitive dependencies regardless of module source language.

serhii-lekariev committed:
Rename a method param and refactor handling `null` values.

serhii-lekariev committed:
Merge remote-tracking branch 'origin/pure-js-boilerplate-elimination' into pure-js-boilerplate-elimination

serhii-lekariev committed:
Fix wrong method call.

serhii-lekariev committed:
Update configuration names to the relevant version exposed by `base`.

serhii-lekariev committed:
Move a method call to a more appropriate place.

Note, that this is the case of a temporal dependency: calling
`disableTransitiveProtos` before a protobuf plugin has been applied
leads to no action, it should always be called afterwards.

serhii-lekariev commented:
@dmdashenkov, PTAL again.

dmdashenkov reviewed (approved):
@serhii-lekariev, LGTM.

serhii-lekariev committed:
Add transitive protos generation tests.

serhii-lekariev committed:
Add a utility test method to enable both Java and JS code gen.
